trie/triedb/pathdb/history.go

```diff
@@ -0,0 +1,594 @@
+	)
+	if err := dec.decode(accountData, storageData, accountIndexes, storageIndexes); err != nil {
+		return nil, err
+	}
+	return &dec, nil
+}
+
+// writeStateHistory writes the state history with provided state set. After
+// storing the corresponding state history, it will also prune the stale
+// histories from the disk with the given threshold.
+func writeStateHistory(freezer *rawdb.ResettableFreezer, dl *diffLayer, limit uint64) error {
+	// Short circuit if state set is not available.
+	if dl.states == nil {
+		return errors.New("state change set is not available")
+	}
+	var (
+		err   error
+		n     int
+		start = time.Now()
+		h     = newHistory(dl.rootHash, dl.parent().root(), dl.blockNumber, dl.states)
+	)
+	accountData, storageData, accountIndex, storageIndex := h.encode()
+	dataSize := common.StorageSize(len(accountData) + len(storageData))
+	indexSize := common.StorageSize(len(accountIndex) + len(storageIndex))
+
+	// Write history data into five freezer table respectively.
+	rawdb.WriteStateHistory(freezer, dl.stateID(), h.meta.encode(), accountIndex, storageIndex, accountData, storageData)
+
+	// Prune stale state histories based on the config.
+	if limit != 0 && dl.stateID() > limit {
+		n, err = truncateFromTail(freezer, dl.stateID()-limit)
+		if err != nil {
+			return err
+		}
+	}
+	historyDataSizeMeter.Mark(int64(dataSize))
+	historyIndexSizeMeter.Mark(int64(indexSize))
+	historyBuildTimeMeter.UpdateSince(start)
+	log.Debug("Stored state history", "id", dl.stateID(), "blockNumber", dl.blockNumber, "data", dataSize, "index", indexSize, "pruned", n, "elapsed", common.PrettyDuration(time.Since(start)))
+	return nil
+}
+
+// checkHistories retrieves a batch of meta objects with the specified range
+// and performs the callback on each item.
+func checkHistories(freezer *rawdb.ResettableFreezer, start, count uint64, check func(*meta) error) error {
+	for count > 0 {
+		number := count
+		if number > 10000 {
+			number = 10000 // split the big read into small chunks
+		}
+		blobs, err := rawdb.ReadStateHistoryMetaList(freezer, start, number)
+		if err != nil {
+			return err
+		}
+		for _, blob := range blobs {
+			var dec meta
+			if err := dec.decode(blob); err != nil {
+				return err
+			}
+			if err := check(&dec); err != nil {
+				return err
+			}
+		}
+		count -= uint64(len(blobs))
+		start += uint64(len(blobs))
+	}
+	return nil
+}
+
+// truncateFromHead removes the extra state histories from the head with the given
+// parameters. It returns the number of items removed from the head.
+func truncateFromHead(freezer *rawdb.ResettableFreezer, nhead uint64) (int, error) {
+	ohead, err := freezer.TruncateHead(nhead)
+	if err != nil {
+		return 0, err
+	}
+	return int(ohead - nhead), nil
+}
+
+// truncateFromTail removes the extra state histories from the tail with the given
+// parameters. It returns the number of items removed from the tail.
+func truncateFromTail(freezer *rawdb.ResettableFreezer, ntail uint64) (int, error) {
```

Comment:
Is there really a legitimate reason to wrap freezer.TruncateTail in truncateFromTail (same for head)? The only extra operation this does it returns the number of items removed that we use for logging in 2 locations, but we could just as well calculate it where it's needed.

IMO it makes things a bit more complicated to have this wrappers + extra error checks that are still needed just the same at the callsites.

Reply:
I still need to add the logic for deleting corresponding root->id mappings. So the wrapper is necessary.

At the beginning, I am hesitated because I am not sure if it's worthwhile. The size of each mapping is 40B,
17M blocks mean 680MB in total, at the worst case(full sync from genesis). 

But I think it's still better to remove them when we truncate history from tail. These mappings are useless anyway.